consensus/safety-rules/src/safety_rules.rs

```diff
@@ -189,11 +189,43 @@ impl<T: Payload> TSafetyRules<T> for SafetyRules<T> {
         ))
     }
 
-    /// @TODO only sign a timeout if it matches last_voted_round or last_voted_round + 1
-    /// @TODO update last_voted_round
+    /// Only sign the timeout if it is greater than or equal to the last_voted_round and ahead of
+    /// the preferred_round. We may end up signing timeouts for rounds without first signing votes
+    /// if we have received QCs but not proposals. Always map the last_voted_round to the last
+    /// signed timeout to prevent equivocation. We can sign the last_voted_round timeout multiple
+    /// times by requiring that the underlying signing scheme provides deterministic signatures.
     fn sign_timeout(&mut self, timeout: &Timeout) -> Result<Ed25519Signature, Error> {
+        debug!("Incoming timeout message for round {}", timeout.round());
+        COUNTERS.requested_sign_timeout.inc();
+
+        let expected_epoch = self.persistent_storage.epoch()?;
+        if timeout.epoch() != expected_epoch {
+            return Err(Error::IncorrectEpoch(timeout.epoch(), expected_epoch));
+        }
+
+        let preferred_round = self.persistent_storage.preferred_round()?;
```

Comment:
this doesn't seem necessary but probably not hurt

Reply:
here's the rub, we don't update last_voted_round if we get a QC from a newer round... so let's assume we were taken offline, see a future QC, now our vote round is at k but our preferred round is k + 10. An adversary could get us to do some weird vote signing. But you're probably right in that it isn't likely strictly required.

Reply:
it's fine to vote on lower round than preferred round as long as this validator never vote on it, but I don't see any problem avoiding it

Reply:
Sure, there's nothing inherently wrong with it. It just isn't a particularly desirable state.